### PR TITLE
All compile, need to implement convexHull

### PR DESCRIPTION
**Pull request by Ixw123**
All compiling now, need to implement convexHull now didn’t realize I was assigned to that as well, i will have questions about that on Monday or later today for you because im not sure which aspect in the packages is the measurement one. That is unless of course you didnt already implement it. This is the PR that matters as its the most up to date.

### CODE REVIEW

**Comment by hudsonb** [comment]
Looks like this PR replaces your previous one? https://github.com/hudsonb/turf.kt/pull/37

If so, can you close that PR.

In the future, it's good practice to have a separate branch and PR per issue. Makes reviews easier.

File: src/main/kotlin/geojsonkt/BBox.kt [hunk 1]
```diff
@@ -1,5 +1,13 @@
 package geojsonkt
 
+import kotlin.Double.Companion.POSITIVE_INFINITY
+import kotlin.Double.Companion.NEGATIVE_INFINITY
+
+//@Suppress("DIVISION_BY_ZERO")
+//const val positiveInfinity: Double = 1.0 / 0.0
+//@Suppress("DIVISION_BY_ZERO")
+//const val negativeINfinity: Double = -1.0 / 0.0
+
```
Comment (hudsonb) [inline src/main/kotlin/geojsonkt/BBox.kt:10]: Remove commented out code.

File: src/main/kotlin/geojsonkt/BBox.kt [hunk 2]
```diff
@@ -47,3 +55,42 @@ fun BBox.toPolygon(): Polygon {
     return Polygon(arrayOf(arrayOf(northWest, southWest, southEast, northEast, northWest)))
 }
 
+// Added bbox things
```
Comment (hudsonb) [inline src/main/kotlin/geojsonkt/BBox.kt:58]: Remove comment

File: src/main/kotlin/geojsonkt/BBox.kt [hunk 3]
```diff
@@ -47,3 +55,42 @@ fun BBox.toPolygon(): Polygon {
...
+
+fun GeoJson.bbox() = {
+    when(this) {
+        is Point -> throw UnsupportedOperationException("Can not calculate BBox of Point Geometry type: ${this::class.java.name}")
+        is LineString -> bbox(coords(this))
+        is Polygon -> bbox(coords(this))
+        is MultiPoint -> throw UnsupportedOperationException("Can not calculate BBox of MultiPoint Geometry type: ${this::class.java.name}")
```
Comment (hudsonb) [inline src/main/kotlin/geojsonkt/BBox.kt:65]: Add bbox support for this

File: src/main/kotlin/geojsonkt/BBox.kt [hunk 4]
```diff
@@ -47,3 +55,42 @@ fun BBox.toPolygon(): Polygon {
...
+        is GeometryCollection -> bbox(coords(this))
+        else -> throw UnsupportedOperationException("Can not calculate BBox of unrecognized Geometry type: ${this::class.java.name}")
+    }
+}
+
+private fun bbox(coords: Sequence<Position>): BBox {
+    var result = doubleArrayOf(POSITIVE_INFINITY, POSITIVE_INFINITY, NEGATIVE_INFINITY, NEGATIVE_INFINITY)
+
+    for(coord in coords) {
+            // Need to look at this and figure out what to do to be able to set these values properly
```
Comment (hudsonb) [inline src/main/kotlin/geojsonkt/BBox.kt:77]: Remove comment

File: src/main/kotlin/turfkt/CenterOfMass.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        var sArea = 0.0
+        var pi: Position
+        var pj: Position
+        var xi: Double
+        var xj: Double
+        var yi: Double
+        var yj: Double
+        var a: Double
+        var translation = centre
+        // val point = coords(this)
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/CenterOfMass.kt:24]: Remove

File: src/main/kotlin/turfkt/Envelope.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import geojsonkt.*
+
+fun getEnvelope(g: Geometry): Polygon = when(g) {
+    is Polygon, 
+    is MultiPolygon,
+    is LineString,
+    is MultiLineString -> g.bbox!!.toPolygon()
+
+    is Point,
+    is MultiPoint -> throw UnsupportedOperationException("Can not BBox Point or Multipoint Geometry type: ${g::class.java.name}")
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/Envelope.kt:12]: Add support for this

File: src/main/kotlin/turfkt/Midpoint.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package turfkt
+
+import geojsonkt.*
+import turfkt.*
+
+fun Position.midpoint(to: Position): Position {
+    val dist = this.distance(to)
+    val heading = this.bearing(to)
+    val units = "" // not sure what to have this default to but i noticed it was not used at all in this instance
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/Midpoint.kt:9]: Remove

File: src/main/kotlin/turfkt/Midpoint.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    return this.destination(dist/2, heading, units)
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/Midpoint.kt:11]: units is optional, just leave it out

File: src/main/kotlin/turfkt/Midpoint.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+fun Point.midpoint(to: Position): Position {
+    val dist = this.distance(to)
+    val heading = this.bearing(to)
+    val units = ""
+
+    return this.destination(dist/2, heading, units)
+}
+// Not sure if you meant a feature of points or not but i think that is all that is implemented
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/Midpoint.kt:21]: Remove

File: src/main/kotlin/turfkt/Midpoint.kt [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+fun Feature<Point>.midpoint(to: Position): Position {
+    val dist = this.distance(to)
+    val heading = this.bearing(to)
+    val units = ""
+
+    return this.destination(dist/2, heading, units)
+}
```
Comment (hudsonb) [inline src/main/kotlin/turfkt/Midpoint.kt:28]: midpoint extensions should all defer to Position.midpoint.

**Comment by hudsonb** [comment]
Envelope is not complete. It looks like currently all it does is return the bbox of the geometry as a `Polygon`, which is really already supported by  the `toPolygon` extension on `BBox`.

According to turf.js docs, envelope:

> Takes any number of features and returns a rectangular Polygon that encompasses all vertices.

So in Envelope.kt we should define an extension on FeatureCollection/GeometryCollection which will calculate the envelope given all of the contained geometries.

File: src/main/kotlin/turfkt/Midpoint.kt (hunk 7 above)
Comment (Ixw123) [reply src/main/kotlin/turfkt/Midpoint.kt:9]: Like the entire function?

Comment (Ixw123) [reply src/main/kotlin/turfkt/Midpoint.kt:9]: Oh nvm you mean the units stuff

File: src/main/kotlin/turfkt/Midpoint.kt (hunk 10 above)
Comment (Ixw123) [reply src/main/kotlin/turfkt/Midpoint.kt:28]: So essentially i just need the one midpoint extension lol

File: src/main/kotlin/turfkt/Midpoint.kt (hunk 8 above)
Comment (Ixw123) [reply src/main/kotlin/turfkt/Midpoint.kt:11]: I get errors when i remove units

**Pull request closed without merging**